File: src-tauri/src/models/foreshadow.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct ForeshadowThreadDto {
    pub id: String,
    pub project_id: String,
    pub title: String,
    pub description: String,
    pub status: String,
    pub setup_note_id: Option<String>,
    pub setup_note_title: Option<String>,
    pub setup_paragraph_id: Option<String>,
    pub payoff_note_id: Option<String>,
    pub payoff_note_title: Option<String>,
    pub payoff_paragraph_id: Option<String>,
    pub setup_document_id: Option<String>,
    pub setup_document_title: Option<String>,
    pub payoff_document_id: Option<String>,
    pub payoff_document_title: Option<String>,
    pub priority: String,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForeshadowThreadInput {
    pub project_id: String,
    pub title: String,
    pub description: Option<String>,
    pub status: Option<String>,
    pub setup_note_id: Option<String>,
    pub payoff_note_id: Option<String>,
    pub setup_document_id: Option<String>,
    pub payoff_document_id: Option<String>,
    pub priority: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateForeshadowThreadInput {
    pub thread_id: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub status: Option<String>,
    pub setup_note_id: Option<Option<String>>,
    pub payoff_note_id: Option<Option<String>>,
    pub setup_document_id: Option<Option<String>>,
    pub payoff_document_id: Option<Option<String>>,
    pub priority: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForeshadowThreadFromNoteInput {
    pub note_id: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub payoff_document_id: Option<String>,
    pub priority: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListForeshadowThreadsInput {
    pub project_id: String,
    pub status: Option<String>,
    pub priority: Option<String>,
    pub only_unpaid: Option<bool>,
}

File: app/types/foreshadow.ts
export type ForeshadowStatus = 'all' | 'open' | 'planned' | 'paid_off' | 'abandoned' | 'archived'
export type ForeshadowPriority = 'all' | 'low' | 'normal' | 'high'
export type ConcreteForeshadowStatus = Exclude<ForeshadowStatus, 'all'>
export type ConcreteForeshadowPriority = Exclude<ForeshadowPriority, 'all'>

export interface ForeshadowThread {
    id: string
    projectId: string
    title: string
    description: string
    status: ConcreteForeshadowStatus | string
    setupNoteId?: string | null
    setupNoteTitle?: string | null
    payoffNoteId?: string | null
    payoffNoteTitle?: string | null
    setupDocumentId?: string | null
    setupDocumentTitle?: string | null
    payoffDocumentId?: string | null
    payoffDocumentTitle?: string | null
    priority: ConcreteForeshadowPriority | string
    createdAt: number
    updatedAt: number
}

export interface CreateForeshadowThreadInput {
    projectId: string
    title: string
    description?: string | null
    status?: ConcreteForeshadowStatus | string | null
    setupNoteId?: string | null
    payoffNoteId?: string | null
    setupDocumentId?: string | null
    payoffDocumentId?: string | null
    priority?: ConcreteForeshadowPriority | string | null
}

export interface UpdateForeshadowThreadInput {
    threadId: string
    title?: string
    description?: string
    status?: ConcreteForeshadowStatus | string
    setupNoteId?: string | null
    payoffNoteId?: string | null
    setupDocumentId?: string | null
    payoffDocumentId?: string | null
    priority?: ConcreteForeshadowPriority | string
}

export interface CreateForeshadowThreadFromNoteInput {
    noteId: string
    title?: string | null
    description?: string | null
    payoffDocumentId?: string | null
    priority?: ConcreteForeshadowPriority | string | null
}

export interface ListForeshadowThreadsInput {
    projectId: string
    status?: ForeshadowStatus | string | null
    priority?: ForeshadowPriority | string | null
    onlyUnpaid?: boolean | null
}

export const foreshadowStatusOptions: Array<{ value: ForeshadowStatus, label: string }> = [
    {value: 'all', label: '全部'},
    {value: 'open', label: '未回收'},
    {value: 'planned', label: '计划回收'},
    {value: 'paid_off', label: '已回收'},
    {value: 'abandoned', label: '放弃'},
    {value: 'archived', label: '归档'}
]

export const foreshadowPriorityOptions: Array<{ value: ForeshadowPriority, label: string }> = [
    {value: 'all', label: '全部'},
    {value: 'high', label: '高'},
    {value: 'normal', label: '普通'},
    {value: 'low', label: '低'}
]

export function foreshadowStatusLabel(status: string) {
    switch (status) {
        case 'open':
            return '未回收'
        case 'planned':
            return '计划回收'
        case 'paid_off':
            return '已回收'
        case 'abandoned':
            return '放弃'
        case 'archived':
            return '归档'
        default:
            return status
    }
}

export function foreshadowPriorityLabel(priority: string) {
    switch (priority) {
        case 'high':
            return '高'
        case 'low':
            return '低'
        default:
            return '普通'
    }
}
